chronos/asyncfutures2.nim

```diff
@@ -217,24 +227,29 @@ proc cancel(future: FutureBase, loc: ptr SrcLoc): bool =
     # mechanism and/or use a regular `addCallback`
     when chronosStrictFutureAccess:
       doAssert future.internalCancelcb.isNil,
-        "futures returned from `{.async.}` functions must not use `cancelCallback`"
-
-    if cancel(future.internalChild, getSrcLocation()):
-      return true
+        "futures returned from `{.async.}` functions must not use " &
+        "`cancelCallback`"
 
+    cancel(future.internalChild, loc)
   else:
-    if not(isNil(future.internalCancelcb)):
+    if isNil(future.internalCancelcb):
+      cancelAndSchedule(future, loc)
+      future.cancelled()
+    else:
       future.internalCancelcb(cast[pointer](future))
       future.internalCancelcb = nil
-    cancelAndSchedule(future, getSrcLocation())
-
-  future.internalMustCancel = true
-  return true
+      if FutureFlag.OwnCancelSchedule notin future.internalFlags:
```

Comment:
instead of a flag, could the `cancelcb` return a boolean?

The code is a bit confusing because both `OwnCancelSchedule` and an empty cancelcb is set for this to work - we could perhaps avoid the state and associated complexity by making this a local decision in the cb.

we can emulate the old behavior in the `cancelcb=` function by wrapping it

Reply:
`cancelcb` could not return a flag, because it will introduce a lot of breaking changes for all projects which has Future[T] closure procedures with cancellation callbacks.

Reply:
> breaking changes

as mentioned, we can maybe work around this with a wrapper, ie
```
proc cancelCb(fut: auto, cb: OldCbType) =
  fut.internalcancelcb = proc(): bool = 
    cb()
    true
```

as is, it's .. difficult to write correct callbacks

Reply:
this would also go in the 4.0 branch so it's a good opportunity

Reply:
I do not like and i will not agree on any changes with callbacks, having multiple callback types just complicate problem solving. So now i prefer Future flags in here, because it
1. Extensible, i can always add/remove flags at any moment without breaking old code. While introducing new callbacks requires deprecation path and so on. More dirty solution as for me.
2. Almost not visible for end users of library.

Reply:
It would be hard to explain why this `result` is needed and what it achieves. And probably we will see a lot of `ugly` results in the code.